pkg/internal/util.bzl

```diff
@@ -0,0 +1,59 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Internal utilities for rules_pkg."""
+
+load(":providers.bzl", "PackageArtifactInfo", "PackageVariablesInfo")
+
+def setup_output_files(ctx, package_file_name=None):
+    """Handle processing of out and package_file_name.
+
+    By default, output is to the "out" attribute. If package_file_name is
+    given, we will write to that name instead, and out will be a symlink to it.
+
+    Callers should:
+       - write to ouput_file
+       - add outputs to their returned DefaultInfo(files) provider
+       - return a PackageArtifactInfo provider of the form:
+            label: ctx.label.name
+            file_name: output_name
+
+    Args:
+      ctx: context
+      package_file_name: computed value for package_file_name
+
+    Returns:
```

Comment:
Perhaps this should say "tuple of ...", or perhaps the return value should be a `struct`?

`struct`s don't support destructuring assignment like with tuples, but I think I'd be fine with either.